pkg/controller/workload/job/job_controller_test.go

```diff
@@ -0,0 +1,149 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"testing"
+
+	batchv1 "k8s.io/api/batch/v1"
+
+	"sigs.k8s.io/kueue/pkg/util/pointer"
+)
+
+func TestJobPodsReady(t *testing.T) {
+	testcases := map[string]struct {
+		job  *batchv1.Job
+		want bool
+	}{
+		"parallelism = completions; no progress": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism: pointer.Int32(3),
+					Completions: pointer.Int32(3),
+				},
+				Status: batchv1.JobStatus{},
+			},
+			want: false,
+		},
+		"parallelism = completions; not enough progress": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism: pointer.Int32(3),
+					Completions: pointer.Int32(3),
+				},
+				Status: batchv1.JobStatus{
+					Ready:     pointer.Int32(1),
+					Succeeded: 1,
+				},
+			},
+			want: false,
+		},
+		"parallelism = completions; all ready": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism: pointer.Int32(3),
+					Completions: pointer.Int32(3),
+				},
+				Status: batchv1.JobStatus{
+					Ready:     pointer.Int32(3),
+					Succeeded: 0,
+				},
+			},
+			want: true,
+		},
+		"parallelism = completions; some ready, some succeeded": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism: pointer.Int32(3),
+					Completions: pointer.Int32(3),
+				},
+				Status: batchv1.JobStatus{
+					Ready:     pointer.Int32(2),
+					Succeeded: 1,
+				},
+			},
+			want: true,
+		},
+		"parallelism = completions; all succeeded": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism: pointer.Int32(3),
+					Completions: pointer.Int32(3),
+				},
+				Status: batchv1.JobStatus{
+					Succeeded: 3,
+				},
+			},
+			want: true,
+		},
+		"parallelism < completions; reaching parallelism is enough": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism: pointer.Int32(2),
+					Completions: pointer.Int32(3),
+				},
+				Status: batchv1.JobStatus{
+					Ready: pointer.Int32(2),
+				},
+			},
+			want: true,
+		},
+		"parallelism > completions; reaching completions is enough": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism: pointer.Int32(3),
+					Completions: pointer.Int32(2),
+				},
+				Status: batchv1.JobStatus{
+					Ready: pointer.Int32(2),
+				},
+			},
+			want: true,
+		},
+		"parallelism specified only; not enough progress": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism: pointer.Int32(3),
+				},
+				Status: batchv1.JobStatus{
+					Ready: pointer.Int32(2),
+				},
+			},
+			want: false,
+		},
+		"parallelism specified only; all ready": {
+			job: &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism: pointer.Int32(3),
+				},
+				Status: batchv1.JobStatus{
+					Ready: pointer.Int32(3),
+				},
+			},
+			want: true,
+		},
+	}
+
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			got := jobPodsReady(tc.job)
```

Comment:
We already have all the coverage in integration.

Reply:
Not everything, for example in integration tests we only test parallelism=completions. Further, the integration tests are much slower, so I would rather decrease the number of scenarios in integration tests than deleting unit tests.